New optimizer manual entry

eneuscamman opened the pull request:
The initial manual entry for the new optimizer is ready.  This update also includes a fix for a possible memory leak in the new optimizer.

ye-luo commented on manual/opt.tex:
```diff
@@ -153,9 +153,108 @@ \subsubsection{quartic}
+   %&   \texttt{stepsize} &  real     & 0--1 & 0.25  & Step size for moving parameters\\
+   &   \texttt{max\_relative\_change} &  real     & $>0$ & 10.0 & Allowed change in cost function\\
+   &   \texttt{max\_param\_change} &  real     & $>0$ & 0.3 & Allowed change in wave function parameter\\
+   &   \texttt{shift\_i} &  real     & $>0$ & 0.01 & Direct stabilizer added to the Hamiltonian matrix\\
```
Change "Direct" to "Initial". OSO doesn't update shift_i but adaptive does. So the user input is an initial value.

eneuscamman commented:
I've just updated the stabilizer shift descriptions based on Ye's comment.

ye-luo commented:
+1

markdewing commented on manual/opt.tex:
```diff
@@ -153,9 +153,108 @@ \subsubsection{quartic}
 
-Manual is needed here.
+The default setting of the adaptive optimizer is to construct the linear method Hamiltonian and overlap matrices explicitly and add different shifts to the Hamiltonian matrix 
+as ``stablizers''. The generalized eigenvalue problem is solved for each shifted Hamiltonian and updates to wave function parameters are obtainted. Then a correlated sampling 
```
'obtained'

markdewing replied:
It might read better as:
The generalized eigenvalue problem is solved for each shift to obtain updates to the wave function parameters.

markdewing commented on manual/opt.tex:
```diff
@@ -153,9 +153,108 @@ \subsubsection{quartic}
-Manual is needed here.
+The default setting of the adaptive optimizer is to construct the linear method Hamiltonian and overlap matrices explicitly and add different shifts to the Hamiltonian matrix 
+as ``stablizers''. The generalized eigenvalue problem is solved for each shifted Hamiltonian and updates to wave function parameters are obtainted. Then a correlated sampling 
+is performed (using the middle shift's updated wave function as the guiding function) for each shift's updated wave function and the initial trial wave function.
```
move the parenthetical comment to the end of the sentence.

eneuscamman commented:
I've updated the text base don Mark's suggestions.

markdewing commented on manual/opt.tex:
```diff
@@ -153,9 +153,108 @@ \subsubsection{quartic}
+The default setting of the adaptive optimizer is to construct the linear method Hamiltonian and overlap matrices explicitly and add different shifts to the Hamiltonian matrix 
+as ``stabilizers''. The generalized eigenvalue problem is solved for each shifted Hamiltonian and updates to wave function parameters are obtained. Then a correlated sampling 
+is performed (using the middle shift's updated wave function as the guiding function) for each shift's updated wave function and the initial trial wave function.
+The cost function for these wave functions is compared, and the update corresponding to the best cost function is selected.
```
I think the comparison of the cost function for the proposed updates can be implicitly assumed.
"The update with the best cost function is selected"

eneuscamman commented:
We wrote it as "cost function" to remind readers that it is not necessarily the energy or the variance, but could also be the excited state targeting function.

markdewing commented on manual/opt.tex:
```diff
@@ -153,9 +153,108 @@ \subsubsection{quartic}
+In the next iteration, the median magnitude of the stabilizers is set to that that generated the best update in the current iteration, thus adapting the magnitude of
+the stabilizers automatically.
+
+Thanks to recently developments in excited state variational principles, the adaptive optimizer is also able to optimize individual excited states directly.  
```
'recent'.
IMO, it might be better to skip the first clause, and add a reference to the paper at the end of the sentence.

markdewing commented on manual/opt.tex:
```diff
@@ -153,9 +153,108 @@ \subsubsection{quartic}
+\begin{equation*}
+\Omega[\Psi]=\frac{\left<\Psi|\omega-H|\Psi\right>}{\left<\Psi|{\left(\omega-H\right)}^2|\Psi\right>}
+\end{equation*}
+The global minimal of this functional corresponds to the state whose energy lies immediately above the shift parameter $\omega$ in the energy spectrum.
```
minimum

markdewing commented on manual/opt.tex:
```diff
@@ -153,9 +153,110 @@ \subsubsection{quartic}
+For example, suppose the initial 
+trial wave function's parameters are taken as a guess for the first excited state.
+Then, by setting $\omega$ to be a value in between the ground state and first excited state 
+energies would cause the trial wave function to be optimized for the first excited state. 
```
energy.  Now I see 'energies' refers to both 'ground state' and 'first excited state'
But it might be a little smoother as:
"Then, by setting $\omega$ to be an energy in between the ground state and the first excited state, the wave function will be optimized for the first excited state."

Another thought - the excited state optimization paragraph might be better moved to it's own subsection, or at least moved to end of the adaptive section.   The description is here because of shared implementation, but the idea of optimizing for excited states is logically separate.

Also - is omega only used for excited state optimization, or is it used for ground state optimization too?
If only for excited states, the bullet point about 'omega' in Additional information, and the recommendation should all be moved to follow the description paragraph.

markdewing commented on manual/opt.tex:
```diff
@@ -153,9 +153,110 @@ \subsubsection{quartic}
+
+When the trial wave function contains more than tens of thousands of parameters, constructing and storing the linear method matrices may become a memory bottleneck. 
+To avoid explicit
+construction of these matrices, the adaptive optimizer implements the recently developed ``block linear method''.  The block linear method (BLM) tries to find an approximate 
```
"recently developed" is going to age poorly in the context of a manual.  I would suggest removing it.

Also the (a), (b), (c) delineations in the next sentence don't seem to add anything (and they don't seem to be referenced later).  Either put the steps in an actual list, or remove them.

The clause following (c) could even be moved to it's own sentence ("Finally, c_opt is estmated by solving ...")

markdewing commented on manual/opt.tex:
```diff
@@ -153,9 +153,110 @@ \subsubsection{quartic}
+
+Additional information:
+\begin{itemize}
+  \item \texttt{shift\_i}.  This is the initial value of direct term added to the diagonal of the Hamiltonian matrix. 
```
"of the direct term"

The last sentence (both after shift_i and shift_s) - do you mean the value of shift_i/s is adjusted by the optimizer as it proceeds?  Is this implicit in describing these as 'the initial value', or is it important to understanding how the algorithm works (ie, that the value the user gives will not be the actual value used later on)

markdewing commented on manual/opt.tex:
```diff
@@ -153,9 +153,110 @@ \subsubsection{quartic}
+                            More stable but slower optimization with a large value. The used value is auto-adjusted by the optimizer. 
+  \item \texttt{shift\_s}.  This is the initial value of the stabilizer based on the overlap matrix added to the Hamiltonian matrix. 
+                            More stable but slower optimization with a large value. The used value is auto-adjusted by the optimizer. 
+  \item \texttt{omega}.     Since approximate wave function has a finite variance of energy, is has been proved that the $\omega$ value
```
"Since the approximate"
"it has been proved"

markdewing commented on manual/opt.tex:
```diff
@@ -153,9 +153,110 @@ \subsubsection{quartic}
+  \item \texttt{nolds}.     In BLM, the inter-block correlation is accounted for by including a small number of wave function update vectors
+                            outside the block. Larger \texttt{nolds} would include more inter-block correlation and more accurate results, but 
+                            also higher memory requirements. 
+  \item \texttt{nkept}.     This is the number of update directions taken from each block in BLM. If all directions are taken in each block, 
```
The parameter name uses 'kept', but the verb in the description is 'taken'.  Might more consistent to use 'kept' in the description.
What setting would be needed to keep all the directions?   Is it problem -dependent?

"fewer than 5 directions"

markdewing commented on manual/opt.tex:
```diff
@@ -153,9 +153,110 @@ \subsubsection{quartic}
+  \item In order to obtain an unbiased excitation energy, one should optimize the ground state with the excited state variational principle as well by setting
+        \texttt{omega} below the ground state energy.  Note that using the ground state variational principle for the ground state and the excited state variational
+        principle for the excited state creates a bias in favor of the ground state. 
+  \item Setting $\omega$ correctly is very important since a too low or too high $\omega$ will lead to a failure of the optimization. We recommend that one use
```
replace "recommend that one use"  with "recommend using" 
"is the approximate"

markdewing commented on manual/opt.tex:
```diff
@@ -153,9 +153,110 @@ \subsubsection{quartic}
+        principle for the excited state creates a bias in favor of the ground state. 
+  \item Setting $\omega$ correctly is very important since a too low or too high $\omega$ will lead to a failure of the optimization. We recommend that one use
+        $\omega=E-\sigma$, in which $E$ is the desired excited state's energy and $\sigma^2$ is your approximate wave function's energy variance.
+  \item The BLM has a higher stochastic prefactor than the standard LM, and so we only advise its use when at least 5,000 variables are being optimized.
```
I assume the stochastic prefactor refers to the run time?

Would the influence on run time be clearer (if not as technically correct) as something like:
The BLM has higher runtime overhead than the standard LM, but better scaling in the number of variables, so we only advise ...

eneuscamman commented:
I've now addressed Mark's suggestions in the above commits.  Especially, I have moved all discussion of excited state targeting to the end of the adaptive section.

prckent commented on manual/opt.tex:
```diff
@@ -153,9 +153,128 @@ \subsubsection{quartic}
+
+When the trial wave function contains more than ten thousand parameters, constructing and storing the linear method matrices may become a memory bottleneck. 
+To avoid explicit construction of these matrices, the adaptive optimizer implements the block linear method (BLM) approach
+(see J. Chem. Theory Comput. DOI: 10.1021/acs.jctc.7b00119).
```
Add reference to bibliography and put citation here.

prckent commented on manual/opt.tex:
```diff
@@ -153,9 +153,128 @@ \subsubsection{quartic}
 %To activate this optimizer, add ``-D BUILD\_LMYENGINE\_INTERFACE=1'' in the CMake command line.
 
-Manual is needed here.
+The adaptive optimizer is also able to optimize individual excited states directly (see  J. Chem. Theory Comput. 12, 3436).  
```
Add to bibliography and put citation here.

eneuscamman commented:
References have been moved to the bibliography as requested.
